File: v131016/TeeNew/Legend_SymbolFlags.pas
unit Legend_SymbolFlags;
{$I TeeDefs.inc}

interface

uses
  {$IFNDEF LINUX}
  Windows, Messages,
  {$ENDIF}
  SysUtils, Classes,
  {$IFDEF CLX}
  QGraphics, QControls, QForms, QDialogs, QExtCtrls, QStdCtrls, QComCtrls,
  QImgList,
  {$ELSE}
  Graphics, Controls, Forms, Dialogs, ExtCtrls, StdCtrls, ComCtrls, ImgList,
  {$ENDIF}
  Base, TeEngine, Series, TeeProcs, Chart, TeCanvas, TeePenDlg;

type
  TLegendSymbolDraw = class(TBaseForm)
    Series1: TPieSeries;
    CheckBox1: TCheckBox;
    ButtonPen1: TButtonPen;
    ImageList1: TImageList;
    procedure FormCreate(Sender: TObject);
    procedure CheckBox1Click(Sender: TObject);
  private
    { Private declarations }
    procedure LegendDraw(Sender: TObject; Series:TChartSeries;
                         ValueIndex:Integer; R:TRect);
  public
    { Public declarations }
  end;

implementation

{$IFNDEF CLX}
{$R *.DFM}
{$ELSE}
{$R *.xfm}
{$ENDIF}

procedure TLegendSymbolDraw.FormCreate(Sender: TObject);
begin
  inherited;

  Series1.Clear;
  Series1.Add(Random(100),'USA');
  Series1.Add(Random(100),'France');
  Series1.Add(Random(100),'Germany');
  Series1.Add(Random(100),'India');
  Series1.Add(Random(100),'Italy');
  Series1.Add(Random(100),'Japan');
  Series1.Add(Random(100),'Portugal');
  Series1.Add(Random(100),'U.K.');

  // Set event to display custom legend symbols:
  Chart1.Legend.Symbol.OnDraw:=LegendDraw;

  ButtonPen1.LinkPen(Chart1.Legend.Symbol.Pen);

  // Make space bigger for flags
  Chart1.Legend.Font.Size:=10;
  Chart1.Legend.VertSpacing:=8;
  Chart1.Legend.Symbol.Width:=30;
  Chart1.Legend.Symbol.WidthUnits:=lcsPixels;
end;

procedure TLegendSymbolDraw.LegendDraw(Sender: TObject; Series:TChartSeries;
                            ValueIndex:Integer; R:TRect);
var Flag : TBitmap;
begin
  if CheckBox1.Checked then
  begin

    Flag:=TBitmap.Create;  // create flag bitmap
    try
      Dec(R.Top,3);
      Inc(R.Bottom,3);

      // copy image from imagelist
      ImageList1.GetBitmap(ValueIndex,Flag);

      with Chart1.Canvas do
      begin
        StretchDraw(R,Flag);  // draw image to legend

        // draw border
        if Chart1.Legend.Symbol.Pen.Visible then
        begin
          Brush.Style:=bsClear;
          AssignVisiblePen(Chart1.Legend.Symbol.Pen);
          Rectangle(R,0);
        end;
      end;

    finally
      Flag.Free;  // destroy bitmap
    end;
  end;
end;

procedure TLegendSymbolDraw.CheckBox1Click(Sender: TObject);
begin
  Chart1.Invalidate;
end;

initialization
  RegisterClass(TLegendSymbolDraw);
end.

File: TeeNew/Builder2007/Legend_SymbolFlags.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "Legend_SymbolFlags.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Base"
#pragma link "TeePenDlg"
#pragma resource "*.dfm"
TLegendSymbolDraw *LegendSymbolDraw;
//---------------------------------------------------------------------------
__fastcall TLegendSymbolDraw::TLegendSymbolDraw(TComponent* Owner)
  : TBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TLegendSymbolDraw::LegendDraw(TObject *Sender,  TChartSeries *Series,
  int ValueIndex, const TRect &R)
{
  TRect bRect = R;
  if (CheckBox1->Checked)
  {
    Graphics::TBitmap *flag = new Graphics::TBitmap(); // create flag bitmap
    try
    {
      bRect.Top -= 3;
      bRect.Bottom += 3;
      // copy image from imagelist
      ImageList1->GetBitmap(ValueIndex,flag);

      Chart1->Canvas->StretchDraw(bRect,flag);  // draw image to legend
      // draw border
      if (Chart1->Legend->Symbol->Pen->Visible)
      {
        Chart1->Canvas->Brush->Style = bsClear;
        Chart1->Canvas->AssignVisiblePen(Chart1->Legend->Symbol->Pen);
        Chart1->Canvas->Rectangle(bRect,0);
      }
    }
    __finally
    {
      delete flag; // destroy bitmap
    }
  }
}
//---------------------------------------------------------------------------

void __fastcall TLegendSymbolDraw::FormCreate(TObject *Sender)
{
  Series1->Clear();
  randomize();
  Series1->Add(random(100),"USA");
  Series1->Add(random(100),"France");
  Series1->Add(random(100),"Germany");
  Series1->Add(random(100),"India");
  Series1->Add(random(100),"Italy");
  Series1->Add(random(100),"Japan");
  Series1->Add(random(100),"Portugal");
  Series1->Add(random(100),"U.K.");

  // Set event to display custom legend symbols:
  Chart1->Legend->Symbol->OnDraw = LegendDraw;

  ButtonPen1->LinkPen(Chart1->Legend->Symbol->Pen);

  // Make space bigger for flags
  Chart1->Legend->Font->Size = 10;
  Chart1->Legend->VertSpacing = 8;
  Chart1->Legend->Symbol->Width = 30;
  Chart1->Legend->Symbol->WidthUnits = lcsPixels;
}
//---------------------------------------------------------------------------

void __fastcall TLegendSymbolDraw::CheckBox1Click(TObject *Sender)
{
  Chart1->Invalidate();  
}
//---------------------------------------------------------------------------

